### ISSUE TITLE
Generalize the AD framework

### ISSUE THREAD

**Issue by smartalecH** [NanoComp/meep#2237]
Currently, meep's python-based adjoint solver parses objective functions and places adjoint sources using `autograd`. There is nice a `Jax` wrapper (thank you, @ianwilliamson!), but it misses out on some important features (like independent gradients of multiple frequencies...although I'm sure we can overcome this someday).

There are quite a few other AD frameworks within the python community alone. It may be worth it to generalize the `OptimizationProblem` API to allow for arbitrary AD packages. This way, developers can more easily incorporate Meep's adjoint solver within their existing tooling.

This doesn't seem too hard. The only thing Meep really needs is a way to automatically compute the scaling factors of each adjoint source from the FOM. Most packages have the equivalent of `autograd`'s `grad` and `jacobian` functions. We could easily build a table of aliases that maps each package's equivalent. If someone is interested in adding support for their own AD flavor, they can just append to this "table" without breaking anything.

From this table, we can build up additional functionality, like a global way to backprop through mapping functions (that is very package dependent right now).

In short, it should be easy to switch AD backends.

**Comment by ianwilliamson** [comment]
> Currently, meep's python-based adjoint solver parses objective functions and places adjoint sources using `autograd`. 

I think you're probably closer than you think. I wouldn't necessarily say that autograd is involved in placing the adjoint sources, since `monitor_values_grad` could be created by converting a tensor from any AD framework (TF, PyTorch, etc) into a numpy array:

https://github.com/NanoComp/meep/blob/0a5aaeea16eeaea50ebeaaee5be8bf34a806adc0/python/adjoint/utils.py#L222-L250

> It may be worth it to generalize the `OptimizationProblem` API to allow for arbitrary AD packages. **This way, developers can more easily incorporate Meep's adjoint solver within their existing tooling.**

VJPs are the bread and butter of all AD frameworks, so I think the most general approach is one which exposes hooks for performing a forward simulation and calculating VJPs w.r.t. that simulation's monitor outputs, i.e. running the adjoint simulation for the requested vector in the VJP.

We pretty much already have these hooks as part of the the existing JAX wrapper, with the JAX specific parts being the use of `custom_vjp` and some of the type hints:

https://github.com/NanoComp/meep/blob/0a5aaeea16eeaea50ebeaaee5be8bf34a806adc0/python/adjoint/wrapper.py#L222-L248

> but it misses out on some important features (like independent gradients of multiple frequencies...although I'm sure we can overcome this someday)

Yeah, this is true. The blocker is the particular way in which JAX has implemented `jax.jacrev()` (reverse-mode Jacobian). It relies on `vmap`, which requires that all underlying operations are JAX primitives with defined batching rules.

I have not thought through all of the details, but I think a solution to this would be along the lines of: 
1. Define a function that performs a Meep simulation, given sources, monitors, geometry, etc.
2. Wrap this function in a JAX primitive and define a [batching rule](https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html#batching) for the frequencies of all grad-able DFT monitors. This may mean that the frequencies need to be supplied as a common JAX array. We may also need the ability to _optionally_ supply frequency-dependent source amplitudes as a JAX array.
3. Define a `jax.custom_vjp` that uses the JAX primitive defined above for both its forward and reverse functions; the reverse pass would use the gradable monitor outputs as the input to the source amplitudes of the JAX primitive

The batching rule of the primitive inside of the custom VJP would allow `jax.value_and_grad` and `jax.jacrev` to be called on any loss function (scalar or vector-valued, respectively) with the `custom_vjp` embedded inside of it.

Roughly speaking, the flow above suggests the need for an "abstraction layer" of sorts around Meep that makes it appear as a frequency domain simulator, with the ability to (optionally) specify frequency-dependent source amplitudes.
